File: src/main/java/io/github/binaryfoo/TagInfo.java
package io.github.binaryfoo;

import io.github.binaryfoo.decoders.*;

/**
 * How a single tag should be decoded.
 */
public class TagInfo {
    private final String shortName;
	private final String longName;
	private final Decoder decoder;
    private final PrimitiveDecoder primitiveDecoder;

    public TagInfo(String shortName, String longName, Decoder decoder) {
        this(shortName, longName, decoder, PrimitiveDecoder.HEX);
    }
    
    TagInfo(String shortName, String longName, Decoder decoder, PrimitiveDecoder primitiveDecoder) {
        this.shortName = shortName;
		this.longName = longName;
		this.decoder = decoder;
        this.primitiveDecoder = primitiveDecoder;
    }

	public String getShortName() {
		return shortName;
	}

	public String getLongName() {
		return longName;
	}

    /**
     * Decode the value of a BerTLv (primitive or constructed) to be included in the children field of a DecodedData.
     *
     * Idea being children are shown in a tree structure.
     */
    public Decoder getDecoder() {
		return decoder;
	}

	public int getMaxLength() {
		return decoder.getMaxLength();
	}

    /**
     * Decode the value of a PrimitiveBerTLv to be included in the decodedData field of a DecodedData.
     *
     * Idea being the decodedData is shown on the same line (same node in the tree) instead of nesting one level down.
     */
    public String decodePrimitiveTlvValue(String valueAsHexString) {
        return primitiveDecoder.decode(valueAsHexString);
    }

}

File: src/main/java/io/github/binaryfoo/TagInfo.kt
package io.github.binaryfoo

import io.github.binaryfoo.decoders.*
import kotlin.collections.mapOf
import kotlin.text.isEmpty

/**
 * How a single tag should be decoded.
 */
data class TagInfo(
        val shortName: String,
        val longName: String,
        val decoder: Decoder, // Decode the value of a BerTlv (primitive or constructed) to be included in the children field of a DecodedData.
        val primitiveDecoder: PrimitiveDecoder = PrimitiveDecoder.HEX,
        val shortBackground: String? = null,
        val longBackground: String? = null) {

    val backgroundReading: Map<String, String?>?
    get() {
        if (shortBackground != null || longBackground != null) {
            return mapOf("short" to shortBackground, "long" to longBackground)
        } else {
            return null
        }
    }

    val fullName: String
    get() {
        return if (longName.isEmpty() || longName == shortName) shortName else shortName + " - " + longName
    }

    fun getMaxLength(): Int {
        return decoder.getMaxLength()
    }

    /**
     * Decode the value of a PrimitiveBerTLv to be included in the decodedData field of a DecodedData.
     *
     * Idea being the decodedData is shown on the same line (same node in the tree) instead of nesting one level down.
     */
    fun decodePrimitiveTlvValue(valueAsHexString: String): String {
        return if (valueAsHexString.isNullOrBlank()) "" else primitiveDecoder.decode(valueAsHexString)
    }

    companion object {

        @JvmStatic fun treeStructured(shortName: String, longName: String, decoder: Decoder, shortBackground: String? = null, longBackground: String? = null): TagInfo {
            return TagInfo(shortName, longName, decoder, PrimitiveDecoder.HEX, shortBackground, longBackground)
        }

    }
}
